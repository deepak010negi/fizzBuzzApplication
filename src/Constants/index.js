export const PROVIDE_ALL_VALUES = 'All fields are mandatory';