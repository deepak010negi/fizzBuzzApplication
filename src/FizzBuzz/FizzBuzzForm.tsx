import React, { useState, useCallback } from "react";
import { PROVIDE_ALL_VALUES } from "../Constants";
import { FormValues } from "../Types";

interface FormProps {
  initialFormValues: FormValues;
  onFormSubmit: (arg0: FormValues) => void;
}

export const useFizzBuzzForm = (props: FormProps) => {
  const { initialFormValues, onFormSubmit } = props;

  const [formValues, setFormValues] = useState(initialFormValues);
  const [submitError, setSubmitError] = useState("");

  const setFormValue = useCallback(
    (event: React.FormEvent<HTMLInputElement>) => {
      const target = event.target as HTMLTextAreaElement;
      const { name, value } = target;

      setFormValues({
        ...formValues,
        [name]: +value
      });
    },
    [formValues, setFormValues]
  );

  const handleSubmitForm = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      const { fizz, buzz, listLength } = formValues;

      let error = !(fizz && buzz && listLength) ? PROVIDE_ALL_VALUES : "";
      setSubmitError(error);
      onFormSubmit(formValues);
    },
    [formValues, setSubmitError, onFormSubmit]
  );

  return {
    setFormValue,
    handleSubmitForm,
    formValues,
    submitError
  };
};

const FizzBuzzForm = (props: FormProps) => {
  const {
    formValues,
    submitError,
    handleSubmitForm,
    setFormValue
  } = useFizzBuzzForm(props);

  return (
    <form onSubmit={handleSubmitForm}>
      <div data-test="input" className="input">
        <div>Fizz Number *:</div>
        <input
          type="number"
          name="fizz"
          min={0}
          onChange={setFormValue}
          value={formValues.fizz}
        />
      </div>
      <div data-test="input" className="input">
        <div>Buzz Number *:</div>
        <input
          type="number"
          name="buzz"
          min={0}
          onChange={setFormValue}
          value={formValues.buzz}
        />
      </div>
      <div data-test="input" className="input">
        <div>List Length Number *:</div>
        <input
          type="number"
          name="listLength"
          min={0}
          onChange={setFormValue}
          value={formValues.listLength}
        />
      </div>
      <p className="note">
        * Please enter values greater than 0 and press submit to see the
        FizzBuzz magic!
      </p>
      <button data-test="submitButton" type="submit" className="submitButton">
        Submit
      </button>
      <div className={submitError ? "errorTextSubmit" : ""}>{submitError}</div>
    </form>
  );
};

export default FizzBuzzForm;
