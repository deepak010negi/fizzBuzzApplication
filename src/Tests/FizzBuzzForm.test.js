import React from "react";
import { shallow, mount } from "enzyme";

import FizzBuzzForm, { useFizzBuzzForm } from "../FizzBuzz/FizzBuzzForm";
import { PROVIDE_ALL_VALUES } from "../Constants";

it("form renders properly", () => {
  const props = {
    initialFormValues: {
      fizz: 3,
      buzz: 5,
      listLength: 15
    },
    onFormSubmit: jest.fn()
  };
  const component = mount(<FizzBuzzForm {...props} />);

  expect(component.html()).toMatchSnapshot();
});

describe("testing useFizzBuzzForm", () => {
  const onFormSubmit = jest.fn();
  const log = jest.fn();
  const props = {
    initialFormValues: {
      fizz: 3,
      buzz: 5,
      listLength: 15
    },
    onFormSubmit
  };

  const Component = props => {
    const hookProps = useFizzBuzzForm(props);
    log(hookProps);

    return null;
  };

  const getHookProps = props => {
    const wrapper = shallow(<Component {...props} />);

    const hookProps = log.mock.calls[0][0];

    const update = () => {
      wrapper.update();

      return log.mock.calls[log.mock.calls.length - 1][0];
    };

    return { hookProps, update };
  };

  beforeEach(() => {
    log.mockClear();
    onFormSubmit.mockClear();
  });

  it("should return correct shape", () => {
    expect(getHookProps(props).hookProps).toMatchSnapshot();
  });

  it("setFormValue should update the formValues object with the name and value from the event object", () => {
    const formValues = {
      fizz: 3,
      buzz: 5,
      listLength: 15
    };

    const { hookProps, update } = getHookProps({
      initialFormValues: formValues,
      onFormSubmit
    });

    expect(hookProps.formValues).toMatchObject(formValues);

    const mockEventObject = {
      target: {
        name: "fizz",
        value: 10
      }
    };
    hookProps.setFormValue(mockEventObject);
    const newHookProps = update();

    expect(newHookProps.formValues.fizz).toBe(10);
  });

  it("handleSubmitForm should call preventDefault on the submit event", () => {
    const preventDefault = jest.fn();
    const mockEventObject = {
      preventDefault
    };

    const { hookProps } = getHookProps(props);

    hookProps.handleSubmitForm(mockEventObject);

    expect(preventDefault).toHaveBeenCalled();
  });

  it("handleSubmitForm should set error if any of the values are falsy", () => {
    const formValues = {
      fizz: 0,
      buzz: 5,
      listLength: 15
    };

    const { hookProps, update } = getHookProps({
      initialFormValues: formValues,
      onFormSubmit
    });

    hookProps.handleSubmitForm({ preventDefault: () => {} });

    expect(update().submitError).toBe(PROVIDE_ALL_VALUES);

    const { hookProps: newHookProps } = getHookProps({
      initialFormValues: {
        fizz: 3,
        buzz: 5,
        listLength: 15
      },
      onFormSubmit
    });

    expect(newHookProps.submitError).toBe("");
  });

  it("handleSubmitForm should call onFormSubmit with the formValues object", () => {
    const formValues = {
      fizz: 10,
      buzz: 5,
      listLength: 15
    };

    const { hookProps } = getHookProps({
      initialFormValues: formValues,
      onFormSubmit
    });

    hookProps.handleSubmitForm({ preventDefault: () => {} });

    expect(onFormSubmit).toHaveBeenCalledWith(formValues);
  });
});

