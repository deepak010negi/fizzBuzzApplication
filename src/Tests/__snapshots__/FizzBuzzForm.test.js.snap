// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form renders properly 1`] = `"<form><div data-test=\\"input\\" class=\\"input\\"><div>Fizz Number *:</div><input type=\\"number\\" name=\\"fizz\\" min=\\"0\\" value=\\"3\\"></div><div data-test=\\"input\\" class=\\"input\\"><div>Buzz Number *:</div><input type=\\"number\\" name=\\"buzz\\" min=\\"0\\" value=\\"5\\"></div><div data-test=\\"input\\" class=\\"input\\"><div>List Length Number *:</div><input type=\\"number\\" name=\\"listLength\\" min=\\"0\\" value=\\"15\\"></div><p class=\\"note\\">* Please enter values greater than 0 and press submit to see the FizzBuzz magic!</p><button data-test=\\"submitButton\\" type=\\"submit\\" class=\\"submitButton\\">Submit</button><div class=\\"\\"></div></form>"`;

exports[`testing useFizzBuzzForm should return correct shape 1`] = `
Object {
  "formValues": Object {
    "buzz": 5,
    "fizz": 3,
    "listLength": 15,
  },
  "handleSubmitForm": [Function],
  "setFormValue": [Function],
  "submitError": "",
}
`;
